File: src/1448.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tree.h"

int dfs(struct TreeNode* root, int maxValue) {
    int retVal = 0;

    if (root == NULL) {
        return retVal;
    }

    if (root->val >= maxValue) {
        retVal++;
        maxValue = root->val;
    }
    int left = dfs(root->left, maxValue);
    int right = dfs(root->right, maxValue);
    retVal += (left + right);

    return retVal;
}
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int goodNodes(struct TreeNode* root) {
    int retVal = 0;

    int maxValue = INT_MIN;  // Each node's value is between [-10^4, 10^4].
    retVal += dfs(root, maxValue);

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
    } testCase[] = {{{3, 1, 4, 3, INT_MIN, INT_MIN, 1, 5}, 8}, {{3, 3, INT_MIN, 4, 2}, 5}, {{1}, 1}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: root = [3,1,4,3,null,1,5]
     *  Output: 4
     *
     *  Input: root = [3,3,null,4,2]
     *  Output: 3
     *
     *  Input: root = [1]
     *  Output: 1
     */

    struct TreeNode* pRoot = NULL;
    int answer = 0;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: root = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            if (testCase[i].nums[j] == INT_MIN) {
                printf("%snull", (j == 0 ? "" : ","));
            } else if (testCase[i].nums[j] == INT_MAX) {
                continue;
            } else {
                printf("%s%d", (j == 0 ? "" : ","), testCase[i].nums[j]);
            }
        }
        printf("]\n");
        pRoot = buildBinaryTree(testCase[i].nums, testCase[i].numsSize, 0);

        answer = goodNodes(pRoot);
        printf("Output: %d\n", answer);

        printf("\n");

        freeTreeNode(pRoot);
        pRoot = NULL;
    }

    return EXIT_SUCCESS;
}

File: src/1448.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop

from tree import TreeNode


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root: TreeNode, maxValue: int) -> int:
        retVal = 0

        if root is None:
            return retVal

        if root.val >= maxValue:
            retVal += 1
            maxValue = root.val

        left = self.dfs(root.left, maxValue)
        right = self.dfs(root.right, maxValue)
        retVal += (left + right)

        return retVal

    def goodNodes(self, root: TreeNode) -> int:
        retVal = 0

        maxValue = float('-inf')  # Each node's value is between [-10^4, 10^4].
        retVal += self.dfs(root, maxValue)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for root in [[3, 1, 4, 3, None, None, 1, 5], [3, 3, None, 4, 2], [1]]:
            # /* Example
            #  *  Input: root = [3,1,4,3,null,1,5]
            #  *  Output: 4
            #  *
            #  *  Input: root = [3,3,null,4,2]
            #  *  Output: 3
            #  *
            #  *  Input: root = [1]
            #  *  Output: 1
            #  */
            logging.info("Input: root = %s", root)

            retVal = pSolution.goodNodes(TreeNode().buildTree(root))
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
